[docs] Uplift `TreeView` customization examples (#11424)
mui/mui-x: docs/data/tree-view/overview/overview.md

## Customization

### Custom icons

> You can customize the collapse, expand and end icons of the TreeView.
> 
> You can use icon liraries:

Reviewer: ```suggestion
Use the `defaultCollapseIcon`, `defaultExpandIcon`, and `defaultEndIcon` props to customize the Tree View icons.
```

We can make it more explicit by already mentioning the props that allow me to add custom icons right away!

Also, removed the "You can use icons libraries" because that made me think we'd be using a library that's _not_ the Material Icons one, even though I now get the meaning!

{{"demo": "CustomMaterialIcons.js", "defaultCodeOpen": true}}

Or you can create your own components:

> {{"demo": "CustomIcons.js", "defaultCodeOpen": false}}

Reviewer: As much as I love how this one looks, I wonder if we need it? 😕 Maybe we can consolidate this with the above and have one prop use an icon from a library and another prop to use a custom "locally"-created icon, with the `SvgIcon` component?

> ### Adding borders

Reviewer: ```suggestion
### Connection border
```

Not sure if this is the best title, but felt like we could be clearer with what "borders" means in the context of the TreeView! Open to other suggestions 😬

Author: Connection border**s** maybe? But other than that, I don't have an idea 🤔

> 

Reviewer: ```suggestion
Target the `treeItemClasses.group` class to add connection borders between the Tree View items.
```

Guess we can have a simple description here so readers can understand _how_ they'd add a border even before expanding the code block.

Author: Rephrased a bit, because my approach there is a bit different, but adding an explanation is a good idea :)

{{"demo": "BorderedTreeView.js", "defaultCodeOpen": false}}

> ### Customizing collapse animations

Reviewer: ```suggestion
### Expand/collapse animation
```

I guess we can ditch the "customizing" verb here, given it's already in the Customization section?
Shorter TOC! 😬

Author: What do you think about 'Toggle animations'? Having 'Expand/collapse' like that in the title feels weird to me 🤔

> 

Reviewer: ```suggestion
Use the [Transition](/material-ui/transitions/) component from Material UI to add animations for when the Tree View expands and collapses.
```

Author: Rephrased a bit, as my approach there is a bit different, but I agree that adding a short explanation is a good idea :)

{{"demo": "CustomAnimation.js", "defaultCodeOpen": false}}

### Customizing for complex use cases

> You can customize the Tree View to accomodate more complex use cases:

Reviewer: Maybe this one should be the one for the animation example? We could simplify it a tiny bit but preserve the animation, which is _very_ nice here! I agree with Flavien, though — the "complex use cases" part is a bit unclear; like, unless we specify it in the demo description, I don't know at a glance what makes it complex.

Author: I honestly see some value in keeping the animation example simpler, lighter. This one tends to be bigger. I also wouldn't ditch the idea of showcasing more interesting demos, but I agree that shoving them under 'complex use cases' isn't optimal :/ 

I would love to find a middle ground between keeping the simpler customization examples as a more 'step by step guide' AND have more complex styled `TreeView` examples to showcase what can be done with this component. But I would love suggestions for how to do that 🤔 

@flaviendelangle 

> I like a lot the idea to try to build some very appealing Tree View to show that it can be done with our component, but I would separate them for the more API focused demos (that would always use the same data set and focus only on a specific change like the border or the animation).

Yeah, that's kind of what I was aiming for, but not sure how to 'separate' these two types of demos in a nice, comprehensive way

Reviewer: Yeah, definitely — the "Common examples" section is something we've established on the Joy docs and are now slowly porting over to the Material UI demo pages. But, given the structure of the MUI X components docs is slightly different (i.e., each component getting several pages as opposed to just one), thus having a page dedicated to Customization, maybe we could have a "Basic customizations" and "Complex customizations" H2s?

Reviewer 2: IMHO we should have two `h2` sections on this page (not sure about the names):

### Customization props

Contains demos that present our various APIs to customize. Each demo comes with 2-3 lines that introduce the API (the classic "You can use the XXX prop to do YYY" or any similar wording). Each demo should be as simple as possible and only focus on one particular API.

These demos should focus on the tools people have at their disposal. The actual UI we build with it is just a pretext and not the main information.

I think that right now we need at least 3 demos in this section:

- One on how to set custom icons (your "Custom icons" demo is great)
- One on how to set custom animations (your "Toggle animations" demo is great, I would just renaming it "Custom toggle animation" because I thought that the demo showed how to switch off the animation)
- One on how to set a custom content (I think that this one might require some API improvement because honestly I hate the DX of the "ContentComponent prop" that requires very long demos for very simple use-cases.

### Common examples

Contains demos that are building coherent UIs, often using several APIs together and that focuses on the look and feel of those demos, not on the clarity of each API.

The name of the demos in this section should focus on what we are trying to build, not the tools we use to do it (not "prop XXX" in the title).

At some point, all the demos we demonstrate here should be use-cases we think should be built with this component. If a demo would be better with `RichTreeView` then it should probably not be here.

On this one we clearly have:

- Connection border (it uses the custom icons but also some custom CSS on the component sub-classes)
- Customizing for complex use cases (would rename that one "File explorer" or something similar because I don't think a single demo should be marked as "complex", we will have other complex customization demos in the future)
- Gmail clone (it's the definition of a common use case)
- The 1st demo of "ContentComponent prop" but rebranded as something like "Expand when clicking the icon" 

____

As time fly we might end up with dozens of demos and maybe we will have to separate into two pages.
But for now I don't think that's something we should worry about.

Author: I really like that structure! Makes a lot of sense to me, seems like a very future-proof approach 🎉Thanks!

Reviewer 2: Depending on which X component we are talking about, the balance between the two sections can vary a lot.
On the pickers, we have way more API to document and almost none "Common examples".
But one could say that we should have more :grimacing:
